R/rows.R

```diff
@@ -775,6 +775,27 @@ rows_insert_prep <- function(con, x_name, y, by, lvl = 0) {
   out
 }
 
+rows_auto_copy <- function(x, y, copy) {
+  name <- remote_name(x)
+  x_types <- get_col_types(remote_con(x), name)[colnames(y)]
+  auto_copy(x, y, copy = copy, types = x_types)
+}
+
+get_col_types <- function(con, name) {
+  if (inherits(con, "TestConnection")) {
+    return(NULL)
+  }
+  if (is_null(name)) {
+    return(NULL)
+  }
+  res <- DBI::dbSendQuery(con, paste0("SELECT * FROM ", name))
+  on.exit(DBI::dbClearResult(res))
+  DBI::dbFetch(res, n = 0)
+  out <- DBI::dbColumnInfo(res)
+
+  set_names(out[[".typname"]] %||% out[["type"]], out[["name"]])
```

Comment:
I think `type` should come first (since that's the spec that `dbColumnInfo()` documents.

I feel like it would be a good idea to somehow test that this still works even if `dbColumnInfo()` returns a column with a different name. I have no idea how common this is, but I do worry about the level of variability in the implementation of `dbColumnInfo()`. I think it currently works sort of by coincidence because `get_col_types()` returns `NULL`, then `NULL[colnames(y)]` is still `NULL`, and presumably `auto_copy(types = NULL)` activates some default behaviour.

Reply:
I now had to realise this does not work for other databases 😞  So, the types are only inferred for Postgres and nothing is done for other databases. To make this work properly for other databases requires either fixes in the corresponding db package or custom SQL on our side.
I added a generic so that users can help us and provide such SQL.